classes/output/report_table.php

```
@@ -0,0 +1,180 @@
+        parent::__construct($uniqueid);
+
+        $this->set_attribute('id', 'local_smartmedia_report_table');
+        $this->set_attribute('class', 'generaltable generalbox');
+        $this->show_download_buttons_at(array(TABLE_P_BOTTOM));
+        $this->is_downloading($params->download, 'smartmedia-report');
+        $this->define_baseurl($url);
+        $this->define_columns(array('type', 'format', 'width', 'duration', 'size', 'cost'));
+        $this->define_headers(array(
+            get_string('report:type', 'local_smartmedia'),
+            get_string('report:format', 'local_smartmedia'),
+            get_string('report:resolution', 'local_smartmedia'),
+            get_string('report:duration', 'local_smartmedia'),
+            get_string('report:size', 'local_smartmedia'),
+            get_string('report:transcodecost', 'local_smartmedia')
+        ));
+        $this->currpage = isset($params->page) ? $params->page : 0;
+        $this->pagesize = isset($params->pagesize) ? $params->pagesize : 0;
+        $this->sortable(true);
+        $this->no_sorting('type');
+        $this->no_sorting('format');
+        $this->no_sorting('cost');
+        $this->set_sql('*', '{local_smartmedia_data}', 'TRUE');
+
+    }
+
+    /**
+     * Get content for type column.
+     *
+     * @param \stdClass $row
+     *
+     * @return string html used to display the type field.
+     * @throws \moodle_exception
+     */
+    public function col_type($row) {
+        $metadata = json_decode($row->metadata);
+
+        if (empty($metadata->videostreams)) {
+            if (!empty($metadata->audiostreams)) {
+                $format = get_string('report:typeaudio', 'local_smartmedia');
+            } else {
+                // We should never get here unless there has been an error with ffprobe when running scheduled task.
+                throw new \coding_exception(get_string('report:typeerror', 'local_smartmedia', $row->id));
```

I don't think we should throw an error in the report.
If FFProbe wasn't able to extract the data from the file then, we should not process the file further and not display it in the table.

---

This is a bit of an edge case, but I'm trying to figure the best way, should I override the table_sql methods to remove the row if it has no audiostreams/videostreams, or should I just add a WHERE clause to the SQL so that only rows which have at least one audio/videostram are displayed? 
My thoughts are that this may be so circumstantial that I should just leave the value blank and pass a debugging message, because if FFprobe has failed it's job, other fields may be blank too, so checking this one specifically may be overkill, or not enough if we want to ensure they all have valid data.

---

I would add a where clause. Have a look at https://github.com/catalyst/moodle-local_smartmedia/blob/master/classes/task/extract_metadata.php#L179 you will see what data is stored in the database when an FFprobe analysis fails.  We don't want to display results of files that can't be analyzed.
Also in the future we won't try to transcode or further process files that have failed FFprobe inspection

---

I've added a `WHERE` clause to the table, however I have kept the coding exception in there for clarity, let me know if you want me to remove it.